### PR TITLE
[Keras Ops and Layer] Add keras.ops.rms_norm() and keras.layers.RMSNormalization()

### PR DESCRIPTION
**Pull request by DavidLandup0**
Adds `RMSNorm` class for symbolic calls and `keras.ops.rms_norm` for functional calls. Might be worth turning into a `Layer` as well?

### Why?

RMS Normalization is getting more and more common in modern architectures, such as Flux ([reference](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/layers/modeling/rms_normalization.py#L8)), DeepSeek ([reference](https://github.com/deepseek-ai/DeepSeek-V3/blob/1398800ebfcd49c048737e8b1aae69dee46ffefc/inference/model.py#L267)) and others.

It also exists as a `torch.nn.functional` function:

```
torch.nn.functional.rms_norm()
```

Given the recent prevalence, it would make sense to expose it as a general easily callable function.

### keras.ops.rms_norm()

The implementation is numerically equivalent to the PyTorch implementation:

```
>>> import numpy as np
>>> import torch
>>> import torch.nn.functional as F
>>> from keras import ops

>>> inputs_np = np.random.rand(1, 10)

>>> ops.rms_norm(inputs_np, (10,))
<tf.Tensor: shape=(1, 10), dtype=float64, numpy=
array([[0.35580455, 0.92887886, 0.72103452, 0.57726157, 0.74101633,
        1.62244326, 1.25164641, 1.54813485, 0.57403427, 0.82654585]])>

>>> F.rms_norm(torch.tensor(inputs_np), (10,))
tensor([[0.3558, 0.9289, 0.7210, 0.5773, 0.7410, 1.6224, 1.2516, 1.5481, 0.5740,
         0.8265]], dtype=torch.float64)
```

### keras.layers.RMSNormalization()

```
import numpy as np
import torch
import keras

>>> layer = keras.layers.RMSNormalization(input_dim=10)
>>> print(layer.scale.shape)
(10,)
>>> inputs_np = np.random.rand(1, 10)

>>> torch.nn.functional.rms_norm(torch.tensor(inputs_np), (10,))
tensor([[0.8799, 0.2891, 1.1446, 0.5699, 0.4804, 0.5580, 1.5239, 1.2463, 1.2682,
         1.2171]], dtype=torch.float64)

>>> layer(inputs_np).numpy()
array([[0.8799191 , 0.28913352, 1.1445931 , 0.56988394, 0.48039594,
        0.5579797 , 1.5239018 , 1.2462722 , 1.2681562 , 1.2171472 ]],
      dtype=float32)
```

### CODE REVIEW

File: keras/src/ops/nn.py [hunk 1]
```diff
@@ -2651,3 +2651,68 @@ def dot_product_attention(
...
+        )
+
+
+@keras_export(
+    [
+        "keras.ops.rms_norm",
+        "keras.ops.nn.rms_norm",
+    ]
+)
+def rms_norm(x, scale=1, axis=-1, epsilon=None):
```
Comment (fchollet) [inline keras/src/ops/nn.py:2678]: You can name it `rms_normalization` (no abbreviations is better)

If it's a stateless op, it doesn't need to be a layer. Doing `x = rms_normalization(x)` will work in both a call context and during functional model construction (strictly equivalent to `x = RMSNormalization()(x)`)

The only motivation for having the layer would be UX -- does it provide a UX benefit?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/src/ops/nn.py [hunk 2]
```diff
@@ -2651,3 +2651,68 @@ def dot_product_attention(
...
+        super().__init__()
+        self.axis = axis
+        self.scale = scale
+        self.epsilon = epsilon
+
+    def compute_output_spec(self, x):
+        return KerasTensor(shape=x.shape)
+
+    def call(self, x):
+        return _normalize(
```
Comment (DavidLandup0) [inline keras/src/ops/nn.py:2667]: Should be rms_norm - fix

File: keras/src/ops/nn.py (hunk 1 above)
Comment (fchollet) [reply keras/src/ops/nn.py:2678]: But is the state trainable -- is it a weight?

Comment (DavidLandup0) [reply keras/src/ops/nn.py]: Strictly speaking - yes, in the same way gamma/beta are in LayerNormalization are:

```
>>> import keras
>>> layer = keras.layers.RMSNormalization(input_dim=10)
>>> print(layer.scale.shape)
(10,)
>>> print(layer.scale.trainable)
True
```

And:

```
>>> layer = keras.layers.LayerNormalization(axis=[1, 2, 3])
>>> layer.build([5, 20, 30, 40])
>>> print(layer.beta.shape)
(20, 30, 40)
>>> print(layer.beta.trainable)
True
```

Or did I miss the question? 😓

File: keras/src/layers/normalization/rms_normalization.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    If `scale` is enabled, the layer will scale the normalized outputs via
+    a learnable scaling factor.
+
+    So, with scaling enabled, the normalization equations
+    are as follows:
+
+    Let the intermediate activations for a mini-batch to be the `inputs`.
+
+    ```
```
Comment (fchollet) [inline keras/src/layers/normalization/rms_normalization.py:20]: Add: python

File: keras/src/layers/normalization/rms_normalization.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    >>> layer.build([5, 20, 30, 40])
+    >>> print(layer.scale.shape)
+    (10,)
+    >>> layer(np.random.rand(1, 10)).numpy()
+    array([[0.35098287, 1.0495652 , 1.4645109 , 1.2944688 , 0.31124955,
+            1.2768592 , 1.184331  , 0.17474432, 0.49955517, 1.2428929 ]],
+        dtype=float32)
+
+    Args:
+        input_dim: int. The dimensionality of the input tensor.
```
Comment (fchollet) [inline keras/src/layers/normalization/rms_normalization.py:36]: Please list all arguments

File: keras/src/layers/normalization/rms_normalization.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    """
+
+    def __init__(self, input_dim, axis=-1, epsilon=1e-6, **kwargs):
```
Comment (fchollet) [inline keras/src/layers/normalization/rms_normalization.py:39]: `input_dim` should not exist as a constructor argument. Instead you can read it in `build(self, input_shape)`, and create `scale` there.

**Review by fchollet — COMMENTED** [review]
Got it, let's keep the layer then!

**Comment by mattdangerw** [comment]
Actually we might be a bit duplicated here? We have `LayerNormalization(rms_scaling=True)` which I think is basically the same thing? Though this separate layer version might be cleaner API wise, idk.

Just wanted to flag before we merge though.

**Comment by DavidLandup0** [comment]
> Actually we might be a bit duplicated here? We have `LayerNormalization(rms_scaling=True)` which I think is basically the same thing? Though this separate layer version might be cleaner API wise, idk.
> 
> Just wanted to flag before we merge though.

In principle, similar - but different computation, i.e.:

>>> layer_norm = keras.layers.LayerNormalization(rms_scaling=True)
>>> layer_norm.build((1, 10))
>>> rms_norm = keras.layers.RMSNormalization()
>>> rms_norm.build((1, 10))
>>> torch.nn.functional.rms_norm(torch.tensor(inputs), (10,))
tensor([[0.0000, 0.1873, 0.3746, 0.5620, 0.7493, 0.9366, 1.1239, 1.3112, 1.4985,
         1.6859]])
>>> rms_norm(inputs)
<tf.Tensor: shape=(1, 10), dtype=float32, numpy=
array([[0.        , 0.18731716, 0.37463433, 0.5619515 , 0.74926865,
        0.9365858 , 1.123903  , 1.3112202 , 1.4985373 , 1.6858544 ]],
      dtype=float32)>
>>> layer_norm(inputs)
<tf.Tensor: shape=(1, 10), dtype=float32, numpy=
array([[0.        , 0.34813422, 0.69626844, 1.0444026 , 1.3925369 ,
        1.7406712 , 2.0888052 , 2.4369395 , 2.7850738 , 3.133208  ]],
      dtype=float32)>

**Comment by fchollet** [comment]
Are they *supposed* to be the same thing? What's the difference? If there is a legit difference then we need to document it in one of the layers (or both)

**Comment by DavidLandup0** [comment]
> Are they _supposed_ to be the same thing? What's the difference? If there is a legit difference then we need to document it in one of the layers (or both)

Sorry for the delay. Agreed, I'll add more documentation on the difference. They're not meant to be the same op. RMS scaling comes from https://arxiv.org/pdf/1910.07467

I haven't actually seen rms scaling in LayerNorm before outside of Keras, but it looks like it's:

```
variance = ops.var(inputs, axis=self.axis, keepdims=True)
inv = ops.rsqrt(variance + self.epsilon)
outputs = (inputs * inv * ops.cast(_broadcast(self.gamma), inputs.dtype)
```

While for RMS norm, it's:

```
rrms = rsqrt(mean(square(x), axis=axis, keepdims=True)+ epsilon)
outputs = (x * rrms) * scale
```

So it's variance vs mean of squares.

**Review by fchollet — APPROVED** [review]
LGTM, thank you
